domain/externalcontroller/state/state.go

```diff
@@ -77,6 +77,57 @@ WHERE  ctrl.uuid=$M.id`
 	return &ci, nil
 }
 
+func (st *State) ControllerForModel(
+	ctx context.Context,
+	modelUUID string,
+) (*crossmodel.ControllerInfo, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	q := `
+SELECT 	(ctrl.uuid, alias, ca_cert, address) as &ExternalController.* 
+FROM   	external_controller AS ctrl	
+	LEFT JOIN external_model AS model
+	ON ctrl.uuid = model.controller_uuid
+       	LEFT JOIN external_controller_address AS addrs
+       	ON ctrl.uuid = addrs.controller_uuid
+WHERE  	model.uuid = $M.id`
+	s, err := sqlair.Prepare(q, ExternalController{}, sqlair.M{})
+	if err != nil {
+		return nil, errors.Annotatef(err, "preparing %q", q)
+	}
+
+	var rows []ExternalController
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return errors.Trace(tx.Query(ctx, s, sqlair.M{"id": modelUUID}).GetAll(&rows))
+	}); err != nil {
+		return nil, errors.Annotate(err, "querying external controller for model")
+	}
+
+	if len(rows) == 0 {
+		return nil, errors.NotFoundf("external controller with model %q", modelUUID)
+	}
+
+	// We know that we queried for a single ID, so the first instance
+	// of the controller info fields will be repeated.
+	ci := crossmodel.ControllerInfo{
```

Comment:
Since this is the same for any single controller, we might as well make a generic conversion of data type rows to domain type.

To make the `Controller` method conform, just include the UUID in the selection.

Reply:
done